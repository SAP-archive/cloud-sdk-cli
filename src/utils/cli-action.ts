/*!
 * Copyright (c) 2019 SAP SE or an SAP affiliate company. All rights reserved.
 */
import cli from 'cli-ux';

export async function action(output: string, showSpinner: boolean, promise: Promise<any>) {
  if (showSpinner) {
    cli.action.start(output);
  } else {
    cli.log(`${output}...`);
  }
  await promise;
  if (showSpinner) {
    cli.action.stop();
  }
}
