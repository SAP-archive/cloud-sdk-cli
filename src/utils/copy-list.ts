/*!
 * Copyright (c) 2019 SAP SE or an SAP affiliate company. All rights reserved.
 */

export interface CopyDescriptor {
  sourcePath: string | URL;
  targetFolder: string;
  fileName: string;
}
