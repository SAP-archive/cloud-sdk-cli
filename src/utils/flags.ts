/*!
 * Copyright (c) 2019 SAP SE or an SAP affiliate company. All rights reserved.
 */

import Command from '@oclif/command';
import { OutputFlags } from '@oclif/parser';

type FlagsInput<CommandT extends typeof Command> = CommandT['flags'];

export type Flags<CommandT extends typeof Command> = OutputFlags<NonNullable<FlagsInput<CommandT>>>;
export type RequiredFlagsParam<CommandT extends typeof Command, ParamsT extends keyof Flags<CommandT>> = Pick<Flags<CommandT>, ParamsT>;
export type OptionalFlagsParam<CommandT extends typeof Command, ParamsT extends keyof Flags<CommandT>> = Partial<
  RequiredFlagsParam<CommandT, ParamsT>
>;
export type FlagsParam<
  CommandT extends typeof Command,
  RequiredParamsT extends keyof Flags<CommandT>,
  OptionalParamsT extends keyof Flags<CommandT>
> = RequiredFlagsParam<CommandT, RequiredParamsT> & OptionalFlagsParam<CommandT, OptionalParamsT>;
