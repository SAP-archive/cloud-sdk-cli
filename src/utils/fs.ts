/* Copyright (c) 2020 SAP SE or an SAP affiliate company. All rights reserved. */

import { promisify } from 'util';
import * as fs from 'fs';
import * as rimraf from 'rimraf';

export const access = promisify(fs.access);
export const mkdir = promisify(fs.mkdir);
export const readdir = promisify(fs.readdir);
export const readFile = promisify(fs.readFile);
export const copyFile = promisify(fs.copyFile);
export const writeFile = promisify(fs.writeFile);
export const rm = promisify(rimraf);
