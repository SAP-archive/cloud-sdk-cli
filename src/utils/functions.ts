/*!
 * Copyright (c) 2019 SAP SE or an SAP affiliate company. All rights reserved.
 */

export interface Confirm {
  confirm: (message: string) => Promise<boolean>;
}

function confirm(message: string): Promise<boolean> {
  return Promise.resolve(true);
}

export const testFunctions = {
  confirm
};
