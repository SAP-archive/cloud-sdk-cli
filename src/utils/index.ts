/*!
 * Copyright (c) 2020 SAP SE or an SAP affiliate company. All rights reserved.
 */

export * from './copy-list';
export * from './generate-vdm-util';
export * from './git-ignore';
export * from './jest-config';
export * from './manifest-yaml';
export * from './message-formatter';
export * from './package-json';
export * from './scaffold';
export * from './templates';
export * from './usage-analytics';
export * from './warnings';
