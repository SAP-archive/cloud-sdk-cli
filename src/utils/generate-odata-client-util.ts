/*!
 * Copyright (c) 2020 SAP SE or an SAP affiliate company. All rights reserved.
 */

import { flags } from '@oclif/command';
import { AlphabetLowercase, AlphabetUppercase } from '@oclif/parser/lib/alphabet';
import { IBooleanFlag, IOptionFlag } from '@oclif/parser/lib/flags';
import * as path from 'path';
import { Options } from 'yargs';
import { GeneratorOptionsSDK } from './generator-options';

interface GeneratorOptionCli {
  projectDir: string;
}

type KeysToOptions = {
  [optionName in keyof GeneratorOptionCli]: Options;
};

export const generatorOptionCli: KeysToOptions = {
  projectDir: {
    default: '.',
    describe: 'Path to the folder in which the VDM should be created. The input and output dir are relative to this directory.',
    type: 'string'
  }
};

type AllOptions = GeneratorOptionsSDK & GeneratorOptionCli;

// OClif distinguishes between boolean and string flags. Split the keys to get proper typing
type FilterBooleanKeys<Base> = {
  [Key in keyof Base]: Base[Key] extends boolean ? Key : never;
};

type BoolArgKeys = NonNullable<FilterBooleanKeys<AllOptions>[keyof AllOptions]>;
export type BoolArgType = {
  [optionName in BoolArgKeys]: IBooleanFlag<boolean>;
};

type StringArgKeys = keyof Omit<AllOptions, BoolArgKeys>;
export type StringArgType = {
  [optionName in StringArgKeys]: IOptionFlag<string | undefined>;
};

export type FlagsParsed = {
  [Key in keyof AllOptions]: AllOptions[Key] extends boolean ? boolean : string | undefined;
};

export function toBooleanFlag(yargsBool: Options): IBooleanFlag<boolean> {
  const extendedDescription = `${yargsBool.describe} [default: ${yargsBool.default}].`;
  return flags.boolean({
    char: yargsBool.alias as AlphabetLowercase | AlphabetUppercase,
    description: extendedDescription,
    required: yargsBool.requiresArg,
    default: yargsBool.default,
    allowNo: yargsBool.default
  });
}

export function toStringFlag(yargsString: Options): IOptionFlag<string | undefined> {
  const options: Partial<IOptionFlag<string>> = {
    char: yargsString.alias as AlphabetLowercase | AlphabetUppercase,
    description: yargsString.describe,
    required: yargsString.requiresArg,
    default: yargsString.default
  };

  if (yargsString?.coerce?.name === 'resolve') {
    options.parse = (input: string, context: any): string => {
      let projectDir = '.';
      if (context.argv.includes('--projectDir')) {
        projectDir = context.argv[context.argv.indexOf('--projectDir') + 1];
      }
      return path.resolve(projectDir, input);
    };
  }
  return flags.string(options);
}
