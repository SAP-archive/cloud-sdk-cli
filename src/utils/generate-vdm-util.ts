/*!
 * Copyright (c) 2019 SAP SE or an SAP affiliate company. All rights reserved.
 */

import { flags } from '@oclif/command';
import { AlphabetLowercase, AlphabetUppercase } from '@oclif/parser/lib/alphabet';
import { IBooleanFlag, IOptionFlag } from '@oclif/parser/lib/flags';
import { GeneratorOptions as GeneratorOptionsSDK, generatorOptionsCli as generatorOptionsSDK } from '@sap/cloud-sdk-generator';
import * as path from 'path';
import { Options } from 'yargs';

interface GeneratorOptionCli {
  projectDir: Options;
}

export const generatorOptionCli: GeneratorOptionCli = {
  projectDir: {
    default: '.',
    describe: 'Path to the folder in which the VDM should be created. The input and output dir are relative to this directory.',
    type: 'string'
  }
};

type AllOptions = GeneratorOptionsSDK & GeneratorOptionCli;

// OClif distinguishes between boolean and string flags. Split the keys to get proper typing
type FilterBooleanKeys<Base> = {
  [Key in keyof Base]: Base[Key] extends boolean ? Key : never;
};

type BoolArgKeys = NonNullable<FilterBooleanKeys<AllOptions>[keyof AllOptions]>;
export type BoolArgType = {
  [optionName in BoolArgKeys]: IBooleanFlag<boolean>;
};

type StringArgKeys = keyof Omit<AllOptions, BoolArgKeys>;
export type StringArgType = {
  [optionName in StringArgKeys]: IOptionFlag<string | undefined>;
};

export type FlagsParsed = {
  [Key in keyof AllOptions]: AllOptions[Key] extends boolean ? boolean : string | undefined;
};

export function toGeneratorSDK(cliFlags: FlagsParsed): GeneratorOptionsSDK {
  return Object.entries(cliFlags)
    .filter(([key, value]) => typeof value !== 'undefined' && generatorOptionsSDK.hasOwnProperty(key))
    .reduce((prev, [key, value]) => ({ ...prev, [key]: value }), {}) as GeneratorOptionsSDK;
}

export function toBooleanFlag(yargsBool: Options): IBooleanFlag<boolean> {
  const extendedDescription = `${yargsBool.describe} [default: ${yargsBool.default}].`;
  return flags.boolean({
    char: yargsBool.alias as AlphabetLowercase | AlphabetUppercase,
    description: extendedDescription,
    required: yargsBool.requiresArg,
    default: yargsBool.default,
    allowNo: yargsBool.default
  });
}

export function toStringFlag(yargsString: Options): IOptionFlag<string | undefined> {
  const options: Partial<IOptionFlag<string>> = {
    char: yargsString.alias as AlphabetLowercase | AlphabetUppercase,
    description: yargsString.describe,
    required: yargsString.requiresArg,
    default: yargsString.default
  };

  if (yargsString?.coerce?.name === 'resolve') {
    options.parse = (input: string, context: any): string => {
      let projectDir = '.';
      if (context.argv.includes('--projectDir')) {
        projectDir = context.argv[context.argv.indexOf('--projectDir') + 1];
      }
      return path.resolve(projectDir, input);
    };
  }
  return flags.string(options);
}
