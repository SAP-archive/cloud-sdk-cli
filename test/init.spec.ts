/*!
 * Copyright (c) 2020 SAP SE or an SAP affiliate company. All rights reserved.
 */
const error = jest.fn();
jest.mock('cli-ux', () => {
  // Mocking needs to happen before the command is imported
  const cli = jest.requireActual('cli-ux');
  return {
    ...cli,
    default: {
      ...cli.default,
      error
    }
  };
});
jest.mock('../src/utils/warnings');

import execa = require('execa');
import * as fs from 'fs-extra';
import * as path from 'path';
import Init from '../src/commands/init';
import { getWarnings, recordWarning } from '../src/utils/warnings';
import { getCleanProjectDir, getTestOutputDir } from './test-utils';

const testOutputDir = getTestOutputDir(__filename);
const expressAppDir = path.resolve('test', 'express');
const nestAppDir = path.resolve('test', 'nest');

describe('Init', () => {
  afterAll(() => {
    fs.removeSync(testOutputDir);
  });

  it('should create a new project with the necessary files', async () => {
    const projectDir = getCleanProjectDir(testOutputDir, 'full-init');
    await Init.run([projectDir, '--projectName=testingApp', '--buildScaffold', '--no-analytics']);

    ['.npmrc', 'credentials.json', 'systems.json', 'manifest.yml']
      .map(file => path.resolve(projectDir, file))
      .forEach(path => {
        expect(fs.existsSync(path)).toBe(true);
      });

    const reportsPath = path.resolve(projectDir, 's4hana_pipeline', 'reports');
    expect(fs.readFileSync(path.resolve(projectDir, 'README.md'), { encoding: 'utf8' })).toInclude('SAP Cloud SDK');

    // execute the ci scripts and check if the reports are written
    await execa('npm', ['run', 'ci-backend-unit-test'], { cwd: projectDir, stdio: 'inherit' });
    expect(fs.readdirSync(path.resolve(reportsPath, 'backend-unit')).length).toBeGreaterThan(1);
    expect(fs.readdirSync(path.resolve(reportsPath, 'coverage-reports', 'backend-unit')).length).toBeGreaterThan(1);

    await execa('npm', ['run', 'ci-integration-test'], { cwd: projectDir, stdio: 'inherit' });
    expect(fs.readdirSync(path.resolve(reportsPath, 'backend-integration')).length).toBeGreaterThan(1);
    expect(fs.readdirSync(path.resolve(reportsPath, 'coverage-reports', 'backend-integration')).length).toBeGreaterThan(1);
  }, 240000);

  it('should create a new project with the necessary files when adding cds', async () => {
    const projectDir = getCleanProjectDir(testOutputDir, 'full-init-cds');
    await Init.run([projectDir, '--projectName=testingApp', '--buildScaffold', '--no-analytics', '--addCds']);

    ['.cdsrc.json', 'srv/cat-service.cds', 'db/data-model.cds', 'src/catalogue/catalogue.module.ts']
      .map(file => path.resolve(projectDir, file))
      .forEach(path => {
        expect(fs.existsSync(path)).toBe(true);
      });

    await execa('npm', ['run', 'cds-deploy'], { cwd: projectDir, stdio: 'inherit' });
    expect(fs.existsSync(path.resolve(projectDir, 'testingApp.db'))).toBeTrue();
  }, 240000);

  it('should add necessary files to an existing project', async () => {
    const projectDir = getCleanProjectDir(testOutputDir, 'add-to-existing');
    fs.copySync(expressAppDir, projectDir, { recursive: true });

    await Init.run([projectDir, '--projectName=testingApp', '--startCommand="npm start"', '--no-analytics', '--skipInstall', '--force']);

    ['.npmrc', 'credentials.json', 'systems.json', 'manifest.yml']
      .map(file => path.resolve(projectDir, file))
      .forEach(path => {
        expect(fs.existsSync(path)).toBe(true);
      });
    expect(fs.existsSync(path.resolve(projectDir, 'test'))).toBe(false);
  }, 10000);

  it('should add necessary files to an existing project when adding cds', async () => {
    const projectDir = getCleanProjectDir(testOutputDir, 'add-to-existing');
    fs.copySync(expressAppDir, projectDir, { recursive: true });

    await Init.run([projectDir, '--projectName=testingApp', '--addCds', '--startCommand="npm start"', '--no-analytics', '--skipInstall', '--force']);

    ['.cdsrc.json', 'srv/cat-service.cds', 'db/data-model.cds']
      .map(file => path.resolve(projectDir, file))
      .forEach(path => {
        expect(fs.existsSync(path)).toBe(true);
      });
  }, 10000);

  it('init should detect and fail if there are conflicts', async () => {
    const projectDir = getCleanProjectDir(testOutputDir, 'detect-conflicts');
    fs.copySync(nestAppDir, projectDir, { recursive: true });
    fs.createFileSync(`${projectDir}/.npmrc`);

    await Init.run([projectDir, '--projectName=testingApp', '--startCommand="npm start"', '--skipInstall', '--no-analytics']);

    expect(error).toHaveBeenCalledWith(
      'A file with the name ".npmrc" already exists. If you want to overwrite it, rerun the command with `--force`.',
      { exit: 1 }
    );
  }, 10000);

  it('should add to .gitignore if there is one', async () => {
    const projectDir = getCleanProjectDir(testOutputDir, 'add-to-gitignore');
    fs.copySync(nestAppDir, projectDir, { recursive: true });

    await Init.run([projectDir, '--projectName=testingApp', '--startCommand="npm start"', '--skipInstall', '--no-analytics']);

    const gitignoreEntries = fs
      .readFileSync(`${projectDir}/.gitignore`, 'utf8')
      .split('\n')
      .filter(entry => entry !== '');

    expect(gitignoreEntries).toContain('credentials.json');
    expect(gitignoreEntries).toContain('/s4hana_pipeline');
    expect(gitignoreEntries).toContain('/deployment');
    expect(gitignoreEntries.length).toBeGreaterThan(29);
  }, 10000);

  it('should show a warning if the project is not using git', async () => {
    const projectDir = getCleanProjectDir(testOutputDir, 'warn-on-no-git');

    fs.createFileSync(path.resolve(projectDir, 'package.json'));
    fs.writeFileSync(path.resolve(projectDir, 'package.json'), JSON.stringify({ name: 'project' }), 'utf8');

    await Init.run([projectDir, '--projectName=testingApp', '--startCommand="npm start"', '--skipInstall', '--no-analytics']);

    expect(recordWarning).toHaveBeenCalledWith(
      'No .gitignore file found!',
      'If your project is using a different version control system,',
      'please make sure the following paths are not tracked:',
      '  credentials.json',
      '  /s4hana_pipeline',
      '  /deployment'
    );
    expect(getWarnings).toHaveBeenCalled();
  }, 10000);

  it('should add our scripts and dependencies to the package.json', async () => {
    const projectDir = getCleanProjectDir(testOutputDir, 'add-scripts-and-dependencies');
    fs.createFileSync(path.resolve(projectDir, 'package.json'));
    fs.writeFileSync(path.resolve(projectDir, 'package.json'), JSON.stringify({ name: 'project' }), 'utf8');

    await Init.run([projectDir, '--projectName=testingApp', '--startCommand="npm start"', '--frontendScripts', '--skipInstall', '--no-analytics']);

    const packageJson = JSON.parse(fs.readFileSync(path.resolve(projectDir, 'package.json'), 'utf8'));

    const dependencies = Object.keys(packageJson.dependencies);
    const devDependencies = Object.keys(packageJson.devDependencies);
    const scripts = Object.keys(packageJson.scripts);

    expect(dependencies).toContain('@sap/cloud-sdk-core');
    expect(devDependencies).toContain('@sap/cloud-sdk-test-util');
    ['ci-build', 'ci-package', 'ci-backend-unit-test', 'ci-frontend-unit-test', 'ci-integration-test', 'ci-e2e'].forEach(script =>
      expect(scripts).toContain(script)
    );
  }, 10000);

  it('should add the analytics file', async () => {
    const projectDir = getCleanProjectDir(testOutputDir, 'add-to-gitignore');
    fs.copySync(nestAppDir, projectDir, { recursive: true });

    await Init.run([
      projectDir,
      '--projectName=testingApp',
      '--startCommand="npm start"',
      '--skipInstall',
      '--analytics',
      '--analyticsSalt=SAPCLOUDSDK4LIFE'
    ]);

    expect(JSON.parse(fs.readFileSync(`${projectDir}/sap-cloud-sdk-analytics.json`, 'utf8'))).toEqual({ enabled: true, salt: 'SAPCLOUDSDK4LIFE' });
  }, 10000);

  it('should add a disabled analytics file', async () => {
    const projectDir = getCleanProjectDir(testOutputDir, 'add-to-gitignore');
    fs.copySync(expressAppDir, projectDir, { recursive: true });

    await Init.run([projectDir, '--projectName=testingApp', '--startCommand="npm start"', '--skipInstall', '--no-analytics']);

    expect(JSON.parse(fs.readFileSync(`${projectDir}/sap-cloud-sdk-analytics.json`, 'utf8'))).toEqual({ enabled: false });
  }, 10000);
});
