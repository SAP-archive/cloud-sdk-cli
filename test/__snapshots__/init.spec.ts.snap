// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Init init should detect and fail if there are conflicts 1`] = `[Error: A file with the name ".npmrc" already exists. If you want to overwrite it, rerun the command with \`--force\`.]`;
